lib/statespace_sse.h

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef STATESPACE_SSE_H_
+#define STATESPACE_SSE_H_
+
+#include <xmmintrin.h>
+
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <cstdint>
+#include <functional>
+
+#include "statespace.h"
+#include "util.h"
+
+namespace qsim {
+
+// Routines for state-vector manipulations.
+// State is a vectorized sequence of four real amplitudes followed by four
```

Comment:
This is a bit confusing. Are these the real and imaginary components of the same four amplitudes? How does this handle states with more or less than four amplitudes? In particular, how are states with `num_amplitudes % 4 != 0` handled?

Reply:
Yes, these are the real and imaginary components of the same four amplitudes. `num_amplitudes % 4 != 0` is possible only for one-qubit circuits (`num_amplitudes == 2`). In this case, the memory is still allocated for four amplitudes. Half of the allocated memory is not used explicitly. For `num_amplitudes > 4` (always multiple of four), the sequence of four real components followed by four imaginary components is repeated many times.